File: leetcode/637.average-of-levels-in-binary-tree/go/averageoflevels.go
package averageoflevels

import "container/list"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func averageOfLevels(root *TreeNode) []float64 {
	res := make([]float64, 0)
	queue := list.New()

	if root != nil {
		queue.PushBack(root)
	}

	for queue.Len() > 0 {
		size := queue.Len()

		subtotal := 0.0

		for range size {
			curr := queue.Remove(queue.Front()).(*TreeNode)

			subtotal += float64(curr.Val)

			if curr.Left != nil {
				queue.PushBack(curr.Left)
			}

			if curr.Right != nil {
				queue.PushBack(curr.Right)
			}
		}

		res = append(res, subtotal/float64(size))
	}

	return res
}

File: leetcode/637.average-of-levels-in-binary-tree/ts/averageoflevels.ts
class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}

const averageOfLevels = (root: TreeNode | null): number[] => {
    const res: number[] = []
    const queue: TreeNode[] = []
    let front = 0

    if (root !== null) queue.push(root)

    while (queue.length > front) {
        const size = queue.length - front

        let subtotal = 0
        for (let i = 0; i < size; i++) {
            const curr = queue[front++]
            subtotal += curr.val
            if (curr.left !== null) queue.push(curr.left)
            if (curr.right !== null) queue.push(curr.right)
        }

        res.push(subtotal / size)
    }

    return res
}

export default averageOfLevels
